hack/codegen.go

```diff
@@ -0,0 +1,734 @@
+
+		err := template.Execute(buf, mapping)
+		if err != nil {
+			return err
+		}
+
+		// 10. if update is set
+		if update {
+			isDirOk := files.IsDir(mapping.GoFile)
+			if !isDirOk {
+				err := os.MkdirAll(filepath.Dir(mapping.GoFile), 0755)
+				if err != nil {
+					return err
+				}
+			}
+
+			err := ioutil.WriteFile(mapping.GoFile, buf.Bytes(), 0644)
+			if err != nil {
+				return err
+			}
+
+			featFile := filepath.Base(path)
+			log.Printf(`Please add '"features/%v": %v.FeatureContext' to features map defined in conformance_test.go (order matters)`,
+				featFile, mapping.Package)
+		}
+
+		return nil
+	}
+
+	if len(mapping.NewFunctions) == 0 {
+		return nil
+	}
+
+	// 13. if update is set
+	if update {
+		log.Printf("Updating go file %v...", mapping.GoFile)
+		err := updateGoTestFile(mapping.GoFile, mapping.NewFunctions)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(mapping.NewFunctions) != 0 {
+		return fmt.Errorf("generated code is out of date")
+	}
+
+	return nil
+}
+
+// Function holds the definition of a function in a go file or godog step
+type Function struct {
+	// Name
+	Name string
+	// Expr Regexp to use in godog Step definition
+	Expr string
+	// Args function arguments
+	// k = name of the argument
+	// v = type of the argument
+	Args *orderedmap.OrderedMap
+}
+
+type Mapping struct {
+	Package string
+
+	FeatureFile string
+	Features    []Function
+
+	GoFile        string
+	GoDefinitions []Function
+
+	NewFunctions []Function
+}
+
+// SignatureChange holds information about the definition of a go function
+type SignatureChange struct {
+	Function string
+	Have     string
+	Want     string
+}
+
+var templateFuncs = template.FuncMap{
+	"backticked": func(s string) string {
+		return "`" + s + "`"
+	},
+	"unescape": func(s string) template.HTML {
+		return template.HTML(s)
+	},
+	"argsFromMap": argsFromMap,
+}
+
+const goTemplate = `
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package {{ .Package }}
+
+import (
+	"github.com/cucumber/godog"
+	"github.com/cucumber/messages-go/v10"
+
+	tstate "github.com/kubernetes-sigs/ingress-controller-conformance/test/state"
+	"github.com/kubernetes-sigs/ingress-controller-conformance/test/kubernetes"
+)
+
+var (
+	state *tstate.Scenario
+)
+
+{{ range .NewFunctions }}
+func {{ .Name }}{{ argsFromMap .Args false }} error {
+	return godog.ErrPending
+}
+{{end}}
+
+func FeatureContext(s *godog.Suite) { {{ range .NewFunctions }}
```

Comment:
I would add a blob of comment on this method saying the code here is generated and should not be modified by hand but rather through `make codegen`.

I also noticed whenever a new Step is added to an existing feature.go file, they are appended _after_ this `FeatureContext` func. For uniformity, I think we should move the `func FeatureContext` declaration at the top of the file.

Reply:
done